File: SortVision/public/code/merge/scala/mergeSort.scala
object MergeSort {
  // Time: Best/Avg/Worst O(n log n), Space: O(n)
  def mergeSort(arr: Array[Int]): Array[Int] = {
    if (arr.length <= 1) return arr

    val mid = arr.length / 2
    val left = mergeSort(arr.slice(0, mid))
    val right = mergeSort(arr.slice(mid, arr.length))
    merge(left, right)
  }

  private def merge(left: Array[Int], right: Array[Int]): Array[Int] = {
    val result = new Array[Int](left.length + right.length)
    var i = 0
    var j = 0
    var k = 0

    while (i < left.length && j < right.length) {
      if (left(i) <= right(j)) { result(k) = left(i); i += 1 }
      else { result(k) = right(j); j += 1 }
      k += 1
    }
    while (i < left.length) { result(k) = left(i); i += 1; k += 1 }
    while (j < right.length) { result(k) = right(j); j += 1; k += 1 }

    result
  }

  def main(args: Array[String]): Unit = {
    val arr = Array(64, 25, 12, 22, 11)
    println(mergeSort(arr).mkString(", "))
  }
}

File: SortVision/public/code/merge/swift/mergeSort.swift
import Foundation

// Merge Sort
// Time: Best/Avg/Worst O(n log n), Space: O(n)
func mergeSort(_ arr: [Int]) -> [Int] {
    guard arr.count > 1 else { return arr }

    let mid = arr.count / 2
    let left = mergeSort(Array(arr[..<mid]))
    let right = mergeSort(Array(arr[mid...]))
    return merge(left, right)
}

private func merge(_ left: [Int], _ right: [Int]) -> [Int] {
    var result: [Int] = []
    result.reserveCapacity(left.count + right.count)

    var i = 0, j = 0
    while i < left.count && j < right.count {
        if left[i] <= right[j] {
            result.append(left[i]); i += 1
        } else {
            result.append(right[j]); j += 1
        }
    }
    result.append(contentsOf: left[i...])
    result.append(contentsOf: right[j...])
    return result
}

let arr = [64, 25, 12, 22, 11]
print(mergeSort(arr))
